core/src/main/java/org/geysermc/geyser/command/CommandRegistry.java

```diff
@@ -0,0 +1,285 @@
+import org.incendo.cloud.CommandManager;
+import org.incendo.cloud.execution.ExecutionCoordinator;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.geysermc.geyser.command.GeyserCommand.DEFAULT_ROOT_COMMAND;
+
+/**
+ * Registers all built-in and extension commands to the given Cloud CommandManager.
+ * <p>
+ * Fires {@link GeyserDefineCommandsEvent} upon construction.
+ * <p>
+ * Subscribes to {@link GeyserRegisterPermissionsEvent} upon construction.
+ * A new instance of this class (that registers the same permissions) shouldn't be created until the previous
+ * instance is unsubscribed from the event.
+ */
+public class CommandRegistry implements EventRegistrar {
+
+    private static final String GEYSER_ROOT_PERMISSION = "geyser.command";
+
+    protected final GeyserImpl geyser;
+    private final CommandManager<GeyserCommandSource> cloud;
+
+    /**
+     * Map of Geyser subcommands to their Commands
+     */
+    private final Map<String, GeyserCommand> commands = new Object2ObjectOpenHashMap<>(13);
+
+    /**
+     * Map of Extensions to maps of their subcommands
+     */
+    private final Map<Extension, Map<String, GeyserCommand>> extensionCommands = new Object2ObjectOpenHashMap<>(0);
+
+    /**
+     * Map of root commands (that are for extensions) to Extensions
+     */
+    private final Map<String, Extension> extensionRootCommands = new Object2ObjectOpenHashMap<>(0);
+
+    /**
+     * Map containing only permissions that have been registered with a default value
+     */
+    protected final Map<String, TriState> permissionDefaults = new Object2ObjectOpenHashMap<>(13);
+
+    public CommandRegistry(GeyserImpl geyser, CommandManager<GeyserCommandSource> cloud) {
+        this.geyser = geyser;
+        this.cloud = cloud;
+
+        // register our custom exception handlers
+        ExceptionHandlers.register(cloud);
+
+        // begin command registration
+        HelpCommand help = new HelpCommand(DEFAULT_ROOT_COMMAND, "help", "geyser.commands.help.desc", "geyser.command.help", this.commands);
+        registerBuiltInCommand(help);
+        buildRootCommand(GEYSER_ROOT_PERMISSION, help); // build root and delegate to help
+
+        registerBuiltInCommand(new ListCommand(geyser, "list", "geyser.commands.list.desc", "geyser.command.list"));
+        registerBuiltInCommand(new ReloadCommand(geyser, "reload", "geyser.commands.reload.desc", "geyser.command.reload"));
+        registerBuiltInCommand(new OffhandCommand("offhand", "geyser.commands.offhand.desc", "geyser.command.offhand"));
+        registerBuiltInCommand(new DumpCommand(geyser, "dump", "geyser.commands.dump.desc", "geyser.command.dump"));
+        registerBuiltInCommand(new VersionCommand(geyser, "version", "geyser.commands.version.desc", "geyser.command.version"));
+        registerBuiltInCommand(new SettingsCommand( "settings", "geyser.commands.settings.desc", "geyser.command.settings"));
```

Comment:
```suggestion
        registerBuiltInCommand(new SettingsCommand("settings", "geyser.commands.settings.desc", "geyser.command.settings"));
```

Reply:
IDEA parameter name hints be doing that to me :sob: